File: core/ShortestDistanceToACharacter/ShortestDistanceToACharacter.py
from typing import List


def shortestToChar(s: str, c: str) -> List[int]:
    left, right, length = 0, 1, len(s)
    res = [0] * length
    while right < length:

        while right < length and s[right] != c:
            if s[left] == c:
                res[right] = min(res[right], abs(right - left)) if res[right] != 0 else abs(right - left)
            right += 1

        while length > right > left:
            if res[left] != 0:
                res[left] = min(res[left], abs(right - left))
            elif s[left] != c:
                res[left] = abs(right - left)
            left += 1

        right += 1

    return res


if __name__ == '__main__':
    print(shortestToChar(s = "loveleetcode", c = "e"))

File: core/ShortestDistanceToACharacter/ShortestDistanceToACharacter.java
package com.dbc;

import java.util.Arrays;

public class ShortestDistanceToACharacter {
    public int[] shortestToChar(String s, char c) {
        int left = 0, right = 1;
        int[] res = new int[s.length()];
        Arrays.fill(res, 0);
        while (right < s.length()) {
            while (right < s.length() && s.charAt(right) != c) {
                if (s.charAt(left) == c) res[right] = Math.abs(right - left);
                right++;
            }
            while (right < s.length() && left < right) {
                if (res[left] != 0) res[left] = Math.min(res[left], Math.abs(right - left));
                else if (s.charAt(left) != c) res[left] = Math.abs(right - left);
                left++;
            }
            right++;
        }
        return res;
    }
}
